datafeed: check remote_cluster_client before cluster aliases in start

schase-es opened the pull request:
TransportStartDatafeedAction previously tried to validate remote index cluster names in datafeed jobs, before checking if the local cluster had remote_cluster_client role. Because this role enables retrieval of the remote cluster names, the validation step would always fail with a no-such-cluster exception. This was confusing. This change moves the remote_cluster_client check ahead of cluster name validation, and adds a test.

Closes ES-11841
Closes: #121149

schase-es committed:
datafeed: check remote_cluster_client before cluster aliases in start

TransportStartDatafeedAction previously tried to validate remote index cluster
names in datafeed jobs, before checking if the local cluster had
remote_cluster_client role. Because this role enables retrieval of the remote
cluster names, the validation step would always fail with a no-such-cluster
exception. This was confusing. This change moves the remote_cluster_client check
ahead of cluster name validation, and adds a test.

Closes ES-11841

schase-es commented:
Including David Kyle as the last person who edited these, as an FYI... this area seems to have limited distributed coordination involvement in the past.

nicktindall commented on modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DatafeedRemoteClusterClientIT.java:
```diff
@@ -0,0 +1,196 @@
+                    .*Datafeed \\[test_datafeed\\] is configured with a remote index pattern\\(s\\) \
+                    \\[remote_cluster:remote_index\\] but the current node \\[local_cluster-0\\] \
+                    is not allowed to connect to remote clusters. Please enable node.remote_cluster_client \
+                    for all machine learning nodes and master-eligible nodes.*""", messages[1]));
```
I wonder if `org.elasticsearch.common.regex.Regex#simpleMatch(java.lang.String, java.lang.String, boolean)` might be more forgiving/readable? it just uses wildcard characters rather than regex syntax.

davidkyle replied:
Usually `assertThat(X, containsString(Y))` is good enough 

```assertThat(e.getMessage(), containsString("Datafeed [test_datafeed] is configured with a remote index pattern  [remote_cluster:remote_index] but the current node [local_cluster-0] is not allowed to connect to remote clusters. Please enable node.remote_cluster_client for all machine learning nodes and master-eligible nodes."));```

nicktindall reviewed (approved):
Some minor nits, but this LGTM. Please wait until we've got approval from someone in ML before merging though, because I'm not 100% familiar with this code.

(specific question to someone more familiar with the code would be - is this the best place to do the check? Node roles are immutable, so we could do the check when the job is created, but the master could change to one that does have the role between creation and starting, and then everything would be fine? so perhaps start is the right place?)

davidkyle commented on modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DatafeedRemoteClusterClientIT.java:
```diff
@@ -0,0 +1,196 @@
+    }
+
+    @Override
+    protected String getTestRestCluster() {
+        return localCluster.getHttpAddresses();
+    }
+
+    @Override
+    protected boolean preserveIndicesUponCompletion() {
+        return true;
+    }
+
+    @Override
+    protected boolean preserveClusterUponCompletion() {
+        return true;
+    }
+
+    @Override
+    protected boolean preserveDataStreamsUponCompletion() {
+        return true;
```
Are you sure you want these settings? If another test is added to this class these settings may interfere with the test results.

schase-es replied:
I'm not sure actually -- the test errorred-out without them. This was adapted from the MultiClustersIT test... I'll see if I can learn more about them

schase-es replied:
Oh I see -- that missing return you flagged in TransportStartDatafeedAction was the culprit -- whatever state that put the thread pools caused an error in clean-up.

davidkyle commented on modules/data-streams/src/javaRestTest/java/org/elasticsearch/datastreams/DatafeedRemoteClusterClientIT.java:
```diff
@@ -0,0 +1,196 @@
+ * License v3.0 only", or the "Server Side Public License, v 1".
+ */
+
+package org.elasticsearch.datastreams;
```
This test should really be in the ml qa suite https://github.com/elastic/elasticsearch/tree/main/x-pack/plugin/ml/qa

The existing [multi cluster tests](https://github.com/elastic/elasticsearch/tree/main/x-pack/plugin/ml/qa/multi-cluster-tests-with-security) are yaml tests and difficult to adapt. Please create a new folder `datafeed-multicluster-tests` in `ml/qa` and move this test to that folder. You will need a new `build.gradle` in the same folder to run the test, I'm not certain what should be in the gradle file but I'll help you figure it out.  As a starting point try this

```
apply plugin: 'elasticsearch.internal-java-rest-test'

dependencies {
  testImplementation project(path: ':test:test-clusters')
  javaRestTestImplementation project(path: xpackModule('core'))  
}
```

Assuming you use the suggested folder names run the test with this command:
```
./gradlew :x-pack:plugin:ml:qa:datafeed-multicluster-tests:javaRestTest 
```

davidkyle replied:
Nice test BTW, thanks for adding it

schase-es replied:
Thanks for these notes David -- this is really helpful. I'm very new here and don't know lots about the testing environment.

schase-es committed:
Addressed review feedback:
- changed doc url to link github issue
- added return after continuation send -- oops!
- use junit methods for testing message/exception matching
- moved the test to the xpack ml/qa folder
- added a gradle build file that appears to work
- removed excess override settings from RestTest

schase-es commented on x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportStartDatafeedAction.java:
```diff
@@ -223,6 +223,19 @@ public void onFailure(Exception e) {
+                                clusterService.getNodeName()
+                            )
+                        )
+                    );
```
Good catch!!

Commit:
[CI] Auto commit changes from spotless

schase-es committed:
Merge branch 'main' into ES-11841_remote-cluster-client-role-datafeed-indices

schase-es committed:
Trimmed extra dependencies from gradle

davidkyle reviewed (approved):
LGTM

schase-es commented:
Great -- everything is merged. Thanks Nick as usualy, and David especially for reviewing something from another team.
